Step3 로또(자동) 코드 리뷰 요청드립니다.

dongha-byun opened the pull request:
안녕하세요. 
우아한 테크캠프 Pro 5기 교육생 변동하 입니다.

step3 로또(자동) 미션에 대한 코드 리뷰 요청드리면서 궁금한 점이 있어서 질문 남겨드립니다.
편하실 때 답변 부탁드립니다.

1. 이번 과제가 유독 어렵게 느껴졌습니다. 몇 시간 동안 해놓은 코드들도 마음에 안들어서 두 세번 다 지우고 새로 하고나니 시간이 너무 많이 흘렀던거 같습니다. 과제를 하면서 제가 객체 설계를 잘 못하는것 같다고 느꼈는데 리뷰어님 께서는 객체설계할 때 어떤 방식으로 설계를 해 나가시는지 조언을 듣고 싶습니다. 

2. 이번에 코드를 작성하면서 로또 번호를 생성하는 전략을 interface로 구현해서 나름 OCP를 준수해보려고 노력했는데 좀 난항이었습니다. 자동 같은 경우에는 난수로 번호를 생성하게 하고 validation만 신경쓰면 어렵다고 생각은 안드는데, 수동 같은 경우에는 사용자의 입력값을 받아서 가공하고 그 입력에 대한 validation 까지 고려해야하다보니 자동에서 구현하기 위해 생각했던 인터페이스 메서드와 안맞는다는 느낌을 좀 받았습니다. 그래서 interface 구현체에서 수동인 경우에는 사용자의 입력을 받는 로직도 포함시켰는데 이게 최선은 아닌거같아 리뷰어님의 생각은 어떠신지 궁금합니다.

바쁘신 와중에도 코드리뷰 꼼꼼하게 해주셔서 항상 감사드립니다.
제가 드린 질문은 이번 코드리뷰에 답변으로 주실 필요는 없으시니 너무 부담 안가져 주셨으면 좋겠습니다...!

항상 섬세한 코드리뷰 감사합니다.

dongha-byun committed:
Docs: 로또 (자동) 요구사항 및 기능목록 정리

dongha-byun committed:
Docs: 로또(자동) 기능목록 수정

dongha-byun committed:
Test: 로또 구매금액 class 테스트코드 작성

dongha-byun committed:
Feat: 로또 구매금액 구현코드 작성

dongha-byun committed:
Test: 로또 번호 테스트코드 작성

dongha-byun committed:
Test: 로또 테스트코드 작성
- 구매로또와 당첨로또 비교 관련 테스트 코드 추가

dongha-byun committed:
Feat: 로또번호 class 구현코드 작성

dongha-byun committed:
Feat: 로또 class 구현코드 작성
- 당첨로또와 구매로또 간 비교 로직 추가

dongha-byun committed:
Test: 로또 당첨 결과 테스트코드 작성

dongha-byun committed:
Feat: 로또 당첨결과 class 구현 코드 작성

dongha-byun committed:
Feat: 생성자 추가

dongha-byun committed:
Refactor: 로또 class 생성자 리펙터링

dongha-byun committed:
Refactor: Lotto 클래스 생성 부분 리펙터링

dongha-byun committed:
Test: 로또 당첨금액 테스트 코드 작성

dongha-byun committed:
Test: 로또 수익률 테스트 코드 작성

dongha-byun committed:
Feat: 로또 수익률 class 구현 코드 작성

dongha-byun committed:
Feat: 로또 당첨금액 class 구현코드 작성

dongha-byun committed:
Docs: 기능목록 정의 수정

dongha-byun committed:
Refactor: 테스트코드 리펙터링
- LottoWinningMoney class 의존관계 변경으로 인한 테스트코드 수정
- LottoReturnRate class 생성자 수정으로 인한 테스트코드 수정

dongha-byun committed:
Feat: 입출력 관련 class 구현코드 추가

dongha-byun committed:
Feat: 로또 번호 생성 규칙 추상화 처리

dongha-byun committed:
Test: 서비스 영역 테스트코드 작성

dongha-byun committed:
Feat: 서비스 영역 구현코드 작성

dongha-byun committed:
Refactor: 도메인 영역 코드 리펙터링

dongha-byun committed:
Feat: main 메서드 작성

chae-yh commented on src/main/java/step3/LottoMain.java:
```diff
@@ -0,0 +1,36 @@
+package step3;
```
step3로 패키지명을 정하면, 해당 코드에 대한 히스토리를 모르는 사람들이 코드를 파악하는데 어려움이 있을 것 같아요. 코드의 특징을 잘 추상화한 이름으로 정해주면 어떨까요...?

dongha-byun replied:
거기까지 미처 생각 못했습니다...ㅠ 다음에 커밋할 때 패키지명 수정해서 올리겠습니다!

chae-yh commented on src/main/java/step3/README.md:
```diff
@@ -0,0 +1,43 @@
+
+--------------------------------------------------------
+
+## 2. 기능목록 정리
```
기능목록 정리 좋네요 👍 그리고 질문 주신 것 중에 첫번째에 대해서는 다른 좋은 곳이 보이지 않아 여기서 답변 드리겠습니다...!

`이번 과제가 유독 어렵게 느껴졌습니다. 몇 시간 동안 해놓은 코드들도 마음에 안들어서 두 세번 다 지우고 새로 하고나니 시간이 너무 많이 흘렀던거 같습니다. 과제를 하면서 제가 객체 설계를 잘 못하는것 같다고 느꼈는데 리뷰어님 께서는 객체설계할 때 어떤 방식으로 설계를 해 나가시는지 조언을 듣고 싶습니다.`

우선, 저는 전체적으로 볼때 객체지향적으로 작성이 되었다고 생각합니다. (반대로 그렇지 못한 설계라면, 데이터들을 전부 가져와서 service layer에서 모든 작업을 하는 god class와 수백줄이 넘어가는 method가 나올 것 같습니다 ㅎㅎ) 그래서 너무 걱정하시진 않아도 될 것 같아요,
조영호님의 `객체지향의 사실과 오해`라는 책에서도 적절히 지적하고 있다시피, 객체지향 설계라는게 어떤 수학 공식처럼 정답이 있는 것은 아니라고 생각합니다. 다만 객체지향적으로 설계할 수 있는 여러가지 좋은 도구들만이 있는 것 같아요. 
저도 설계를 할 때는 과정에서 강조하는 `객체지향 생활체조 원칙`을 염두에 두면서 작업을 합니다. 그리고 추가로 객체의 구조가 기존의 디자인패턴 등으로 잘 설명이 될 것 같다면 해당 패턴을 도입하는 식으로 작업을 하곤 합니다. 그러면 제 개인적으로는 자연스럽게 역할이 분리되는 것 같아요. 하지만 저도 부족한 점이 많아서 항상 공부하고 부족한 점을 채워나가려고 노력하고 있습니다 ㅎㅎ

객체 지향 개발관련하여 개인적으로 도움이 되었던 내용을 소개하면서 답변 마무리드리겠습니다...!

`협력에 필요한 책임을 결정하고 객체에게 책임을 할당하는 과정을 얼마나 합리적이고 적절하게 수행했는지가 객체지향 설계의 품질을 결정한다.

객체의 행위에 초점을 맞추기 위해서는 협력이라는 실행 문맥 안에서 책임을 분배해야 한다.`
from 객체지향의 사실과 오해 / 조영호 저

사실 좋은 객체지향 설계라는 내용은 간단한 내용은 아니라서, 여러가지 책들을 참고하시면서 공부하는 것도 좋을 것 같습니다. 이미 유명한 조영호님의 객체지향의 사실과 오해나 오브젝트와 같은 책들을 추천드립니다...!

dongha-byun replied:
좋은 답변 감사드립니다!!!!!

chae-yh commented on src/main/java/step3/domain/Lotto.java:
```diff
@@ -0,0 +1,51 @@
+    public static final int NOT_CONTAIN_VALUE = 0;
+    public static final int NUMBER_COUNT = 6;
+    private List<LottoNumber> numbers;
+    private Lotto(){}
```
코드 템플릿이 적용되지 않은 것 같아요. [다른 교육생분들이 적용한 것](https://nextstep-members.slack.com/archives/C048F1DLU72/p1666930409544319)을 참고하여 적용부탁드립니다...!
별개로, static factory method에서만 객체를 만들기 위해 private 로 막은 것으로 보이는데 좋네요 👍

dongha-byun replied:
앗... 너무 설계에 집중해서 놓친거같습니다...ㅠ 적용하겠습니다...!

chae-yh commented on src/main/java/step3/domain/Lotto.java:
```diff
@@ -0,0 +1,51 @@
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+public class Lotto {
```
일급 컬렉션 👍

chae-yh commented on src/main/java/step3/domain/Lotto.java:
```diff
@@ -0,0 +1,51 @@
+        Lotto lotto = new Lotto();
+        lotto.numbers = numbers.stream()
+                .sorted()
+                .map(number -> new LottoNumber(number))
```
LottoNumber도 static factory method를 만들어 줄 수 있지 않을까요...?

dongha-byun replied:
보다보니 그랬네요...ㅠ 이부분도 static factory method로 풀어보겠습니다!

chae-yh commented on src/main/java/step3/domain/Lotto.java:
```diff
@@ -0,0 +1,51 @@
+        return this.numbers.contains(number);
+    }
+
+    public int containNumberCount(Lotto target) {
```
메서드는 보통 동사형으로 사용하는데, contain으로 시작하니 boolean을 return해야 할 것 같네요 ㅎㅎ
getContainNumberCount나 countIfContainsNumber는 어떨까요..?

dongha-byun replied:
미션 진행할 때는 크게 신경쓰지 않았는데 다시보니 뭔가 어색하네요...ㅎ 적절하게 수정하겠습니다!

chae-yh commented on src/main/java/step3/domain/Lotto.java:
```diff
@@ -0,0 +1,51 @@
+        return lotto;
+    }
+
+    public boolean isContain(LottoNumber number) {
```
해당 메서드는 클래스내에서만 사용하는 듯 싶네요 private으로 막아도 좋을 것 같습니다

dongha-byun replied:
넵! 앞으로 이런 부분도 고려하겠습니다!!

chae-yh commented on src/main/java/step3/domain/LottoNumber.java:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    @Override
+    public int hashCode() {
```
equals 와 hashCode 재정의 👍

chae-yh commented on src/main/java/step3/domain/LottoPayment.java:
```diff
@@ -0,0 +1,24 @@
+
+import java.util.regex.Pattern;
+
+public class LottoPayment {
```
구입 금액에 해당하는 클래스네요 👍

chae-yh commented on src/main/java/step3/domain/LottoWinningMoney.java:
```diff
@@ -0,0 +1,41 @@
+import java.util.List;
+import java.util.Map;
+
+public class LottoWinningMoney {
```
순위에 따른 금액을 정리한 클래스네요 👍 
해당 형태로 구성을 한다면, 열거형(ENUM)으로 구성해주면 더 읽기 쉬울 것 같아요..!

dongha-byun replied:
오... ENUM 생각을 못했네요...ㅠ 한 번 ENUM으로 해보겠습니다!

chae-yh commented on src/main/java/step3/service/LottoService.java:
```diff
@@ -0,0 +1,48 @@
+
+import java.util.List;
+
+public class LottoService {
```
이 부분은 궁금하여 여쭤봅니다. 현재 service layer가 세개로 나뉘었는데, 분리하신 기준은 어떻게 될까요...?

dongha-byun replied:
요구사항을 명시하고 기능목록을 작성해보니 전체적인 process가 3단계 정도로 나눠졌습니다. 1) 로또 구매 및 자동번호 수령 / 2) 당첨번호와 구매번호 비교 / 3) 당첨결과 확인 및 수익률 계산 으로 process가 나눠진다고 생각했습니다. 해당 부분들을 하나의 service에서 모두 해결하기엔 service 가 너무 커졌었고, interface로 구현해놓은 로또번호생성 전략도 1)번 process 에선 자동으로, 2)번 프로세스에선 수동으로 처리했어야돼서 service를 분리하고 각각 다른 구현체로 DI 해줄수 있겠다고 생각해서 service를 나누게 되었습니다...!

chae-yh replied:
보통 layered architecture에서 controller - service - domain 으로 구성이 되어 있다고 할 때 service layer는 

1. DAO 연결
2. 서로 다른 도메인의 데이터를 조합
3. Transactional 혹은 Cache 사용

정도로 생각이 됩니다. 해당 구조로 봤을 때 현재 저희가 다루는 도메인은 Lotto라는 하나의 도메인이고 따라서 service(혹은 controller)도 하나로 되어야 하지 않을까 생각해봤습니다.
동하님 작성해주신 내용을 봤을 때 어떤 순서를 기준으로 분리를 해주신 것 같은데, 순서 정도의 개념이라면 controller에서 작성하고 각각 메서드 분리정도로도 가능하지 않을까 싶은데, 동하님은 어떻게 생각하시는지 궁금합니다 ㅎㅎ

chae-yh commented on src/main/java/step3/service/LottoService.java:
```diff
@@ -0,0 +1,48 @@
+
+public class LottoService {
+
+    private List<Lotto> lottos;
```
lottos의 경우에는 모든 경우 파라미터로 값이 들어오기 때문에 필드(혹은 전역변수)로 사용하지 않아도 될 것 같아요...!

dongha-byun replied:
넵...! 앞으로 이런 상황도 충분히  고민해서 적용해보겠습니다 ㅎㅎ

chae-yh commented on src/main/java/step3/strategy/AutoLottoNumberStrategy.java:
```diff
@@ -0,0 +1,36 @@
+    @Override
+    public List<Integer> generateNumbers() {
+        validateRange(MIN_NUMBER, MAX_NUMBER);
+        validateCount(MIN_NUMBER, MAX_NUMBER, NUMBER_COUNT);
```
검증 이란 것은 모르는 입력값에 대해서 적절한 값인지를 판단하는 것이라고 볼때 validateRange, validateCount의 경우는 이미 알고 있는 상수 값에 대해서 다시 검증을 해주고 있습니다. 해당 작업이 필요할까요..?

dongha-byun replied:
말씀해주시고 나서 다시 보니 필요없는 작업이라는 생각이 드네요...ㅠ 다음부터 더 깊이있게 고민해봐야겠습니다 ㅠ 감사합니다!

chae-yh commented on src/main/java/step3/strategy/LottoNumberStrategy.java:
```diff
@@ -0,0 +1,8 @@
+
+public interface LottoNumberStrategy {
+
+    List<Integer> generateNumbers();
```
인터페이스와 구현체의 구분 👍

chae-yh commented on src/test/java/step3/domain/LottoNumberTest.java:
```diff
@@ -0,0 +1,19 @@
+    @ParameterizedTest
+    @ValueSource(ints = {0, 46})
+    @DisplayName("로또 번호는 1~45 안에 속하지 않으면 예외를 발생시킨다.")
+    void 로또숫자_범위_예외_테스트(int value) {
```
`@DisplayName 으로 테스트에 대해 자세히 설명해줄 수 있다면 메서드명은 영어로 간단하게 작성해보면 어떨까요..?

dongha-byun replied:
앗 넵 ㅎㅎ @DisplayName 을 쓰게되면 메서드명은 간단하게 작성해보겠습니다! ㅎㅎ

chae-yh commented on src/test/java/step3/domain/LottoResultTest.java:
```diff
@@ -0,0 +1,81 @@
+import static org.assertj.core.api.Assertions.*;
+import static org.junit.jupiter.api.Assertions.*;
+
+public class LottoResultTest {
```
꼼꼼한 테스트 👍

chae-yh commented on src/main/java/step3/domain/LottoReturnRate.java:
```diff
@@ -0,0 +1,40 @@
+    }
+    public LottoReturnRate(int winningMoney, int payment) {
+        validatePayment(payment);
+        returnRate = winningMoney / (double)payment ;
```
부동소수점 연산은 가끔 예상하지 못한 결과가 return되는 경우가 있습니다. BigDecimal을 써보는 것은 어떨까요...?

dongha-byun replied:
넵! 수정해보도록 하겠습니다!

chae-yh commented on src/main/java/step3/strategy/ConsoleLottoNumberStrategy.java:
```diff
@@ -0,0 +1,28 @@
+
+import static step3.io.InputUtils.*;
+
+public class ConsoleLottoNumberStrategy implements LottoNumberStrategy{
```
질문주신 내용은 여기서 답변드리겠습니다...!

`이번에 코드를 작성하면서 로또 번호를 생성하는 전략을 interface로 구현해서 나름 OCP를 준수해보려고 노력했는데 좀 난항이었습니다. 자동 같은 경우에는 난수로 번호를 생성하게 하고 validation만 신경쓰면 어렵다고 생각은 안드는데, 수동 같은 경우에는 사용자의 입력값을 받아서 가공하고 그 입력에 대한 validation 까지 고려해야하다보니 자동에서 구현하기 위해 생각했던 인터페이스 메서드와 안맞는다는 느낌을 좀 받았습니다. 그래서 interface 구현체에서 수동인 경우에는 사용자의 입력을 받는 로직도 포함시켰는데 이게 최선은 아닌거같아 리뷰어님의 생각은 어떠신지 궁금합니다.`

우선 LottoNumberStrategy 인터페이스로 추출해내면서 생기는 장점은, AutoLottoNumberStrategy 구현체의 경우 generateNumbers()를 실행할 경우 자동으로 랜덤인 값들의 리스트를 return 시켜주기 때문에 테스트가 어렵습니다. 예를 들어 generateNumbers()를 실행시켰을 때 두개의 똑같은 번호의 로또를 만들어야 하는 테스트가 있다면 AutoLottoNumberStrategy 로는 테스트가 어려울 것으로 보입니다. 반대로, 현재는 해당 부분이 인터페이스로 되어 있기 때문에 지금 만들어주신 형태 등의 구현체로 쉽게 똑같은 번호의 로또를 만들 수 있을 것 같아요.

하지만 작성해주신 것은 사용자의 입력을 받는 부분까지 ConsoleLottoNumberStrategy 에 포함 시켰는데, 해당 구조도 저는 꽤 괜찮다고 생각합니다. 다만, 사용자의 입력을 받는 부분/ 혹은 출력 부분은 view의 역할로, 사용자의 입력에 대해서 바로 Lotto.create()를 사용하는 방법도 있을 것 같아요. ㅎㅎ
사용자의 입력을 받는 부분/ 혹은 출력 부분을 view의 역할로 보는지, 아니면 domain에 포함시켜도 좋을지의 논의가 필요할 것 같습니다만, 현재 상태로도 괜찮을 것 같습니다...!

dongha-byun replied:
좋은 답변 감사드립니다! 저도 사용자의 입출력으로 생성되는 로또번호인지라 이걸 view 역할로 해놓을지 아니면 로또번호 생성에 초점을 맞춰 domain 영역으로 봐야할지에 대한 고민을 여러 번 했습니다.ㅠㅠ 그래도 중요한 데이터의 생성에 초점을 맞춰서 개발하다보니 이게 맞나... 싶었는데 여러 방면으로 좋은 말씀 해주셔서 감사드립니다!!

chae-yh commented on src/main/java/step3/domain/Lotto.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    @Override
+    public String toString() {
```
toSting도 재정의하셨군요 👍

chae-yh commented on src/main/java/step3/domain/LottoNumber.java:
```diff
@@ -0,0 +1,37 @@
+    public static final int MIN_NUMBER = 1;
+    public static final int MAX_NUMBER = 45;
+    private int number;
+    public LottoNumber(int number) {
```
여기서 조금만 더 욕심을 부려보면, 로또 번호는 제한된 숫자만 만들 수 있습니다.
해당 값을 미리 캐시해두고 사용해보면 어떨까요...?

dongha-byun replied:
좋을 것 같습니다...! 한 번 시도해보겠습니다!

chae-yh commented on src/main/java/step3/domain/LottoResult.java:
```diff
@@ -0,0 +1,55 @@
+public class LottoResult {
+    public static final int MIN_CONTAIN_COUNT = 3;
+    public static final int ZERO = 0;
+    private Map<Integer, List<Lotto>> result;
```
Lotto의 리스트가 아니라 그냥 개수만 가지고 있어도 모든 경우를 계산해줄 수 있지 않을까요..?

dongha-byun replied:
이 부분을 좀 많이 고민했는데 어떤 로또가 3개가 맞았는지에 대한 정보를 나중에 쓸 수 도 있겠다는 가정으로 만들어놓은 부분이었습니다 ㅎㅎ 처음 PR 드릴때 List 를 가져갔으니 이번에 수정할 때 개수만 가지고 있도록 수정해보겠습니다!

chae-yh commented on src/main/java/step3/domain/LottoNumber.java:
```diff
@@ -0,0 +1,37 @@
+
+import java.util.Objects;
+
+public class LottoNumber {
```
원시값 포장 👍

chae-yh reviewed (changes requested):
동하님 안녕하세요. 피드백 드리는게 늦었습니다..ㅠ
몇번씩 지우면서 작성해주셨다고 말씀해주셨는데, 그러한 고민이 여러군데서 보이는 코드였습니다...! 그리고 걱정하신 것이 기우로 느껴질만큼 잘 작성하신 것 같아요.
다만, 조금 더 고민해주시면 좋을만한 내용이 있어 코멘트 드렸으니 확인 부탁드리고, 추가로 질문 주신 부분도 답변드려보았는데 확인 부탁드립니다...!

감사합니다 : )

dongha-byun committed:
Refactor: 패키지명 수정
- 패키지명 수정 ( step3 -> lottoauto)

dongha-byun committed:
Refactor: 메서드명 수정
- Lotto class 내 메서드 명 변경 ( containNumberCount -> getContainNumberCount )

dongha-byun committed:
Refactor: 메서드 접근자 수정
- Lotto class 내 isContain 메서드 접근자 수정(public -> private)
- 테스트 코드에서 사용되는 부분이 있지만, 같은 테스트 class 에서 isContain을 포함한 테스트가 별도로 존재해 isContain 단독 테스트 코드는 삭제함

dongha-byun committed:
Refactor: AutoLottoNumberStrategy class 코드 리펙터링
- validate 로직 불필요로 인해 삭제처리

dongha-byun committed:
Refactor: 수익률 계산 시, double 대신 BigDecimal 사용하도록 수정
- LottoReturnRate class 필드 및 생성자 수정
- LottoReturnRate 수정으로 인한 관련 Service 및 Test 코드 수정

dongha-byun committed:
Fix: 당첨 로또 입력 2회 오류 수정

dongha-byun committed:
Refactor: 수익률 계산 시, double 대신 BigDecimal 사용하도록 수정

dongha-byun committed:
Refactor: 로또 당첨결과 저장 시, 당첨 수량만 저장하도록 수정
- LottoResult class 내 result Map 에 List 대신 Integer 로 처리하도록 수정
- Map value 변경으로 인한 관련 로직 및 테스트코드 수정

dongha-byun committed:
Refactor: 수익률 계산 시, BigDecimal 캡슐화
 - LottoReturnRate 생성 시, class 내부에서 BigDecimal로 변환하도록 수정

dongha-byun committed:
Refactor: 당첨금 정보 enum 으로 처리하도록 수정
- LottoWinningMoneyEnum class 생성
- 테스트코드 동시 추가 및 winningMonry 관련 로직 수정

dongha-byun committed:
Refactor: 로또번호 캐싱 처리
- LottoNumber class 에서 1~45 List 캐시생성 추가
- AutoLottoNumberStrategy 에서 LottoNumber의 캐시를 통해 로또번호 생성하도록 수정

dongha-byun committed:
Refactor: Service 영역 코드 리펙터링
- LottoService class 내 필드변수 의존관계 축소

chae-yh commented on src/main/java/lottoauto/LottoMain.java:
```diff
@@ -0,0 +1,37 @@
+package lottoauto;
```
피드백 반영 좋습니다 👍 다만 현재 패키지로는 수동 로또에 대한 대응이 안될 것 같아요...! 
이 부분은 추후에 확인 부탁드립니다 ㅎㅎ

chae-yh commented on src/main/java/lottoauto/domain/Lotto.java:
```diff
@@ -0,0 +1,49 @@
+    public static final int NOT_CONTAIN_VALUE = 0;
+    public static final int NUMBER_COUNT = 6;
+    private List<LottoNumber> numbers;
+    private Lotto(){}
```
코드 템플릿이 여전히 적용이 안된 것 같습니다..ㅠ

chae-yh commented on src/main/java/lottoauto/domain/LottoNumber.java:
```diff
@@ -0,0 +1,52 @@
+    public static final int MIN_NUMBER = 1;
+    public static final int MAX_NUMBER = 45;
+    private int number;
+    public static final List<Integer> CACHE = new ArrayList<>();
```
피드백 반영 👍

chae-yh commented on src/main/java/lottoauto/domain/LottoWinningMoney.java:
```diff
@@ -0,0 +1,39 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public class LottoWinningMoney {
```
이 클래스는 사용하지 않는 클래스인 것 같은데 삭제해보면 어떨까요..?

chae-yh reviewed (approved):
동하님 안녕하세요. 피드백 전체적으로 잘 반영해주셨습니다 👍 다만, 코드 템플릿은 아직 적용이 안된 것 같아서, 다음 step 진행하실 때는 적용해서 pr부탁드립니다.

해당 내용 제외하면 크게 피드백 드릴 내용은 없어서 이만 merge 하겠습니다. 
다음 step 진행 부탁드립니다. 감사합니다 : )
